x-pack/test/functional/apps/discover/feature_controls/discover_security.ts

```diff
@@ -185,9 +185,14 @@ export default function({ getPageObjects, getService }: FtrProviderContext) {
 
       it(`doesn't show visualize button`, async () => {
         await PageObjects.common.navigateToApp('discover');
-        await setDiscoverTimeRange();
-        await PageObjects.discover.clickFieldListItem('bytes');
-        await PageObjects.discover.expectMissingFieldListItemVisualize('bytes');
+        await retry.try(async () => {
+          await setDiscoverTimeRange();
+          const hasNoResults = await PageObjects.discover.hasNoResults();
```

Comment:
I would generally try to avoid checking for something to not exist since it takes a timeout of 10 seconds or so.  Compared to checking for something that should exist like the hit count.  Don't change anything yet.  I'm going to run these tests locally and see if I have a suggestion for a change.

Reply:
The current code as it retries a couple of times to make sure it's not on the "no results" page (it's not the default timeout, but a 2500ms timeout) takes about 3 seconds;

```
[11:39:34.333150000] │ debg TestSubjects.exists(discoverNoResults)
[11:39:34.393075600] │ debg Find.existsByDisplayedByCssSelector('[data-test-subj="discoverNoResults"]') with timeout=2500
[11:39:34.704008100] │ debg --- retry.tryForTime error: [data-test-subj="discoverNoResults"] is not displayed
[11:39:35.236323000] │ debg --- retry.tryForTime failed again with the same message...
[11:39:35.761950200] │ debg --- retry.tryForTime failed again with the same message...
[11:39:36.293043400] │ debg --- retry.tryForTime failed again with the same message...
[11:39:36.829574800] │ debg --- retry.tryForTime failed again with the same message...
[11:39:37.354057600] │ debg TestSubjects.click(field-bytes)
```

vs getting the hitCount and verifying it's > 0 takes about .2 seconds
```
[11:54:10.954650700] │ debg TestSubjects.getVisibleText(discoverQueryHits)
[11:54:11.002496700] │ debg TestSubjects.find(discoverQueryHits)
[11:54:11.052031400] │ debg Find.findByCssSelector('[data-test-subj="discoverQueryHits"]') with timeout=10000
[11:54:11.113423700] │ debg TestSubjects.click(field-bytes)
```

```
@@ -187,8 +187,9 @@ export default function({ getPageObjects, getService }: FtrProviderContext) {
         await PageObjects.common.navigateToApp('discover');
         await retry.try(async () => {
           await setDiscoverTimeRange();
-          const hasNoResults = await PageObjects.discover.hasNoResults();
-          expect(hasNoResults).to.be(false);
+          const hitCount = await PageObjects.discover.getHitCount();
+          // eslint-disable-next-line radix
+          expect(parseInt(hitCount)).to.be.greaterThan(0);

           await PageObjects.discover.clickFieldListItem('bytes');
           await PageObjects.discover.expectMissingFieldListItemVisualize('bytes');
@@ -281,8 +282,10 @@ export default function({ getPageObjects, getService }: FtrProviderContext) {
         await PageObjects.common.navigateToApp('discover');
         await retry.try(async () => {
           await setDiscoverTimeRange();
-          const hasNoResults = await PageObjects.discover.hasNoResults();
-          expect(hasNoResults).to.be(false);
+          const hitCount = await PageObjects.discover.getHitCount();
+          // eslint-disable-next-line radix
+          expect(parseInt(hitCount)).to.be.greaterThan(0);
+
           await PageObjects.discover.clickFieldListItem('bytes');
           await PageObjects.discover.expectMissingFieldListItemVisualize('bytes');
         });
@@ -362,8 +365,9 @@ export default function({ getPageObjects, getService }: FtrProviderContext) {
         await PageObjects.common.navigateToApp('discover');
         await retry.try(async () => {
           await setDiscoverTimeRange();
-          const hasNoResults = await PageObjects.discover.hasNoResults();
-          expect(hasNoResults).to.be(false);
+          const hitCount = await PageObjects.discover.getHitCount();
+          // eslint-disable-next-line radix
+          expect(parseInt(hitCount)).to.be.greaterThan(0);
```

Reply:
Sorry but I just realized another potential issue with this change.

In other Discover tests we've used a retry only around getting the hit count and comparing it to the expected value.  We didn't include setting the time range in the retry because each time you set the timepicker it's going to reload the page, and it's the page loading we're waiting for with the retry.

From the failing test issue you said
"the screenshot of the failed test is telling me, no data available, expand your time range. that's odd"

Did the screenshot show the expected start and end dates?

Reply:
thx @LeeDr , back today, I'll soon provide feeback

Reply:
About the screenshot, yes it's showing the defined time range, but no data: 
![image](https://user-images.githubusercontent.com/463851/79334689-0acbd400-7f21-11ea-8fe8-9f712226d44e.png)

Reply:
Here's a flaky suite:
Executions: 42, Failures: 2
https://kibana-ci.elastic.co/job/kibana+flaky-test-suite-runner/337/
It's hard to debug, since the the test result also says "No test failed"
But it was constantly failing until I've implemented the retry. Never failed locally.

Reply:
I've run a similar test suite in OSS for debugging the issue, it's wasn't flaky there:
https://kibana-ci.elastic.co/job/kibana+flaky-test-suite-runner/339/

Reply:
dear @LeeDr, wonder how to proceed here? 

Maybe switch to `const hitCount = await PageObjects.discover.getHitCount(); ` , since this fixes the test, and open another issue because of the flaky data fetching to investigate?

Thx!

Reply:
I find it pretty concerning that the screenshot shows the correct dates in the timepicker and no results?!?!  It *could* still just be a timing issue that the results just haven't come back in the response yet, but the pink loading bar isn't there either so that doesn't feel right.

I'm looking at the flaky-test-suite-runner output now....

FYI, here's an example of a test where we only put the getHitCount() in the retry because it's waiting for the response from Elasticsearch and for the page to load that data;  https://github.com/elastic/kibana/blob/master/test/functional/apps/discover/_discover.js#L74

Reply:
/cc @lukasolson (any other thoughts on this?)

I had a couple of thoughts on debugging this while running the test locally.

We could turn on Elasticsearch slowlogs on both the `logstash-*` and `.async-search` indices.  I don't see that we've done that in any existing tests yet.  It's a per-index setting.  Seems like it would have to be done after `esArchiver.loadIfNeeded('logstash_functional');`.  But the slowlog only shows the query, not the response.  So this might not help in debugging the issue.

Another thing you could try, is if we fail to find hit count, or if we do find the "no results" page, is to try to open the inspector and capture the request and response.  It could show that either the query sent was wrong, or the query was right and Elasticsearch didn't return the correct response, or the correct response was returned and Discover didn't display it.

Or temporarily add debug logging to output the query and response to the Kibana log.